### ISSUE TITLE
Intermittent Very Loud Burst of Noise

### ISSUE THREAD

**Issue by spddy101** [mumble-voip/mumble#4538]
One and sometimes both users of a three-person murmur server that I am hosting hear a very loud burst of noise approximately once every few hours presumably from my connected mumble application or generated by the murmur server (I never hear this burst of noise).  I have not been able to identify anything correlated to the issue.  A key detail is that there are times when both connected users can hear it at the same time and sometimes only one can hear it, seemingly indicating that it isn't an issue with my audio input.

Unknown how to reproduce at will.

I am running Windows 10, Murmur 1.3.2, and Mumble 1.3.2.  The issue was also happening when I was running a 1.3.0 snapshot build.

The user who hears this issue the most often is running Mumble 1.3.0

See/hear the example attached.  This is recorded from the server member who hears this the most often.  
[problem example_Trim_Crop2.zip](https://github.com/mumble-voip/mumble/files/5413079/problem.example_Trim_Crop2.zip)

Please advise on any further troubleshooting steps to take or additional information to provide.

Thank you.

**Comment by Krzmbrzl** [comment]
When this has happened, have you checked the console output? This can also be seen when opening the "developer console" from within Mumble (must be activated in the settings and then accessed from the menu-bar of Mumble)?

Does anything special show up in the server-logs?

I do suspect that this is a client-side issue as the server only relays the audio but doesn't actually touch it.

Are you using positional audio? Or are you talking with one another "normally"? Is there any special channel setup that you are using (ACLs, linked channels, ...)?

**Comment by spddy101** [comment]
I have not checked the Mumble console output.  I (and the users experiencing the sound) will do this next time it occurs.

Nothing comes up in the murmur server logs but I will check this again when it happens.

I do have positional audio enabled in Mumble but it doesn't function in the games I play (Rocket League and Apex Legends).  I don't believe there is any special channel setup (I haven't purposefully set something up).

**Comment by Krzmbrzl** [comment]
Okay then let's see if the logs will show us something useful :)

**Comment by magnus-gross** [comment]
I think it is very interesting that the bug appears as soon as you have rotated your car in such a way that Y was changing the UP and DOWN direction in very short succession. I was having the same issue in GTA V with positional Audio: When I was rotating a jet like you did your car in the video, I was having audio problems too but this was fixed by #4172 

I know you said you have Positional Audio not working, but can you please chech next time when the problem appears, if someone rotated their car like that?

**Comment by sregister** [comment]
My group has heard this too after years of using mumble. To us it was seemingly unexplained (dont think anyone uses positional audio). I always assumed it was a latency or lag issue of some sort. Could the voice data stream lose packets in someway which causes the corruption of the audio?

**Comment by TerryGeng** [comment]
I think a recorded audio clip of such bursts is extremely useful for debugging the problem, so we can know the duration of it and other useful information that help us locate what exactly is wrong. I don't really think it is a network problem since we use some jitter buffer and it smoothes out the network gitch pretty well. You may try the builtin record function in mumble.

**Comment by spddy101** [comment]
@TerryGeng, please see the .zip file in the original post for the example audio clip.

The issue has not come up again yet.  I will be sure to post the Mumble and Murmur logs here when it does.

**Comment by TerryGeng** [comment]
@spddy101 Sorry for not have noticed that...

I investigated the clip. Looks like the problematic region is

![image](https://user-images.githubusercontent.com/2306637/97096393-1d785000-169e-11eb-8561-f796a867e735.png)

It has a duration of 300ms, and it apparently happens when people are talking.
300ms is too long for a single opus packet (the maximum length is 120ms) so I don't think it is related to encoding/decoding.
What is the length of your audio buffer? In most cases, they are ~10ms so I don't think it is a system glitch...

Actually, just by this pattern, I'm suspecting if it is just the jitter buffer causing the problem, triggered by packets loss.

You have mentioned that sometimes everyone hears this but sometimes only one user can hear the glitch. I think the users can immediately examine the "Info" after hearing that glitch to see if there's a significant increase in losing packets.

I can always write a patch to dump audio from every point in our audio process chain, but the raw audio would be huge. They may even stuff your hard disk completely before the glitch happens...

**Comment by spddy101** [comment]
No problem, TerryGeng.  I like the analysis.

All,
Please see the following zip file that contains another audio example, and the mumble logs of myself (User1) and the user experiencing the noise burst (User2) as well as the murmur logs.
[Mumble_Noise_Issue.zip](https://github.com/mumble-voip/mumble/files/5436214/Mumble_Noise_Issue.zip)

No additional items showed up in the logs when the issue happened.  It occurred around 21:20 while the logs appear to only show info from startup.  Note that a third user was connected to the server and he did not hear this burst of noise.

Please let me know how I can further assist this debug.  Thanks!

**Comment by TerryGeng** [comment]
Hi! You can check the packet loss by right-clicking on yourself or other users and hit the "Information".

![image](https://user-images.githubusercontent.com/2306637/97123357-71e70280-1766-11eb-881d-ef1adc35f361.png)

I think it would be very useful if you examine if there's significant packet loss when the burst happens.

Another thing to check is the jitter buffer section in the setting panel:

![image](https://user-images.githubusercontent.com/2306637/97123410-b6729e00-1766-11eb-9455-3a7d80e85e1a.png)

Can you let us know your settings?

**Comment by spddy101** [comment]
Please see the below screenshots with my (User1) settings along with the packet info on my end as well as the packet info for User2 right after User2 experienced the "noise burst".

![Mumble_Setting_User1](https://user-images.githubusercontent.com/71985264/97506879-3b151600-1939-11eb-98e5-005ccf4b85a4.png)
![Mumble_UDP_Stats_User1](https://user-images.githubusercontent.com/71985264/97506883-3c464300-1939-11eb-9153-eaea74fad784.png)
![Mumble_UDP_Stats_User2](https://user-images.githubusercontent.com/71985264/97506882-3c464300-1939-11eb-94a4-7e210fbbecba.png)

**Comment by spddy101** [comment]
Please see the requested information below.  I am User1 and the user experiencing the noise bursts is User2.  The packet info is captured right after User2 hears the noise burst.

User 1
![Mumble_UDP_Stats_User1](https://user-images.githubusercontent.com/71985264/97507186-e1f9b200-1939-11eb-843f-6d36b18a30d4.png)

User 2
![Mumble_UDP_Stats_User2](https://user-images.githubusercontent.com/71985264/97507185-e1f9b200-1939-11eb-90e2-0cf01367619e.png)

User 1 Settings
![Mumble_Setting_User1](https://user-images.githubusercontent.com/71985264/97507184-e1611b80-1939-11eb-92bd-afeededaf9e2.png)

**Comment by spddy101** [comment]
Please see the requested information below.  I am User1 and the user experiencing the noise bursts is User2.  The packet info is captured right after User2 hears the noise burst.

User 1
![Mumble_UDP_Stats_User1](https://user-images.githubusercontent.com/71985264/97507366-4f0d4780-193a-11eb-8569-0846f5d013f2.png)

User 2
![Mumble_UDP_Stats_User2](https://user-images.githubusercontent.com/71985264/97507377-53396500-193a-11eb-88c6-49957a6361d9.png)

User 1 Settings
![Mumble_Setting_User1](https://user-images.githubusercontent.com/71985264/97507388-59c7dc80-193a-11eb-947e-84b064a15ba6.png)

**Comment by spddy101** [comment]
Hi @Krzmbrzl and @TerryGeng,
Is there any other information I can provide that would be helpful in debugging this issue? 

Thanks

**Comment by Krzmbrzl** [comment]
I currently have no idea how this could happen. Therefore I sadly can't let you debug it more since I don't know what to look for.

If you were to find a way to reliably reproduce the issue, that'd make debugging a lot simpler. I guess though that the chances of that happening are rather slim...

**Comment by TerryGeng** [comment]
Sorry for being not able to respond immediately...

It looks like your network is fine, no significant packet loss was shown in the screenshots. I have no clues what is wrong either... I'd suggest you increase the length of the jitter buffer further, perhaps to 20ms or more. I can also write a patch this weekend to log packet loss, or dump all audio data from every node of our audio process chain.

**Comment by spddy101** [comment]
No problem, thanks for checking in on this.  I'll try increasing the jitter buffer.  I don't necessarily know what's involved with running the patch you may develop, TerryGeng, but I can certainly try to use it if you write it.

**Comment by MartB** [comment]
This is the exact same sound we once heard with a noisy microphone user that enabled RNNOISE.
Might be a RNNOISE implementation issue.

**Comment by spddy101** [comment]
The sound burst happened again with RNNoise and positional audio (an earlier suggestion) disabled.  RNNoise has never been enabled by myself or the other user.

**Comment by Green-Sky** [comment]
Those sound burst have happened for all the time i have used mumble, rarely, usually not more than twice a day with no relation to operating system, positional audio and rnnoise. "For all the time i have used mumble" means since ~2014. We actually never got a sample though, but its a signature mumble sound artifact, so don't just remove it /s .

**Comment by Gert-dev** [comment]
Some of my fellow Mumblers have recently updated from 1.3.3 to an unstable version leading to 1.4.0, and started hearing a similar, but not identical, issue where a burst of white noise of about 200 ms (give or take) is sometimes audible.

I myself had been running an unstable version several months before the others, and also had this issue without others updating, but what's strange is: we all hear it at the _exact_ same moment.

That I heard it alone first seemed to imply it's some sort of issue that came into play after 1.3.3, and that it is related to local output rather than input from someone else causing it, but since we now all hear it at the exact same time, implies instead that someone (the server?) is transmitting this to everyone - which suggests that either the server is corrupting something or someone on Mumble <= 1.3.3 is causing it for others > 1.3.3.

**Comment by Krzmbrzl** [comment]
Given that the server does not touch the audio stream (it only relays them), I think it is unlikely that the server is corrupted.

I think a client-side issue is more likely :thinking:

**Comment by MarioPL98** [comment]
I can also confirm the issue. Happends on both LAN and through the Internet. All 1.3.x versions have this issue. Not sure about earlier.

**Comment by Krzmbrzl** [comment]
@MarioPL98 do you have RNNoise noise suppression enabled?

**Comment by MarioPL98** [comment]
Only normal noise suppression. 

One of my friends had huge amount of noises when echo cancellation was enabled in their laptop sound chipset driver (not mumble). Other one uses Cloud II Pro sound a370 prime prodapter and also has noises (one about every 15-60 sec).

I use onboard 1220 on asus prime x370 pro and friends say that I also get noises from time to time.

 I think the issue might be related to certain hardware implementations.

**Comment by iBoonie** [comment]
The ear-piercing scream has been happening to my group since 2013, it started when Opus was added, if that helps at all.

**Comment by Krzmbrzl** [comment]
The latest insights into this issue seem to indicate that this could be related to old clients connecting to the server that don't support Opus. I quote:
> What seemed to trigger it was a client connecting to our server that didn't support opus. We had the default opusthreshold=100 value set in the murmur config, so this caused murmur to switch codecs away from opus. Immediately after this happened, a ton of users started complaining about really loud static.
>
> We have since changed the opusthreshold to 0 and haven't had issues anymore.

So this might be related to switching codec on the server. Could anyone check if setting `opusthreshold = 0` in the server config resolves the issue for them as well?

**Comment by iBoonie** [comment]
Ill preface that all of my clients are running within the last 1 or 2 releases, so everyone is on Opus. We have been running `opusthreshold = 0` since as long as I can remember, bursts still happen.

**Comment by Krzmbrzl** [comment]
I think someone on IRC mentioned that their client wanted to switch to CELT even though they had `opusthreshold = 0` set as well. So maybe this is what is happening to you as well :thinking:

**Comment by TerryGeng** [comment]
@Krzmbrzl IIRC we are going to drop the CELT support in 1.4.0 completely. Is this still our plan now?

**Comment by TerryGeng** [comment]
@Krzmbrzl IIRC we are going to drop the CELT support in 1.4.0 completely. Is this still our plan now?

**Comment by Krzmbrzl** [comment]
Whether it will happen for 1.4.0 depends on whether or not this gets done before 1.4.0 is released. It is not on our list of things that absolutely have to happen before we release.
But yeah: In the future we want to drop it.

**Comment by iBoonie** [comment]
I was going to edit what I wrote and ask if its possible if the client could possibly be (trying to) fall back to CELT, but you beat me to it. Is there a way to force a client to only be on Opus, or would we have to strip it out and recompile to test?

**Comment by Krzmbrzl** [comment]
@davidebeatrici is there already a way to force Opus on the client, regardless of what the server instructs? I think not, right?

Crafting such a version should be rather easy though :thinking:

**Comment by davidebeatrici** [comment]
No way to do it, but easy to implement.

**Comment by Krzmbrzl** [comment]
@iBoonie you can find a modified version that (should) enforces Opus here: https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=3930&view=artifacts&pathAsName=false&type=publishedArtifacts

Please let us know if this happens to fix your issue :)

**Comment by iBoonie** [comment]
@Krzmbrzl I have gotten my group to install it, so I will let you know if it fixes the problem in the coming week or two!

**Comment by Krzmbrzl** [comment]
@iBoonie what's the status on this? :)

**Comment by iBoonie** [comment]
Sorry for the delay, so far so good, but we are going to keep testing for a while since it happens pretty rare. First thing I noticed is the person that the burst of noise happens to the most, his mic sounds more clear, everyone else's sounds the same.

**Comment by Krzmbrzl** [comment]
Well that at least sounds promising :)

**Comment by iBoonie** [comment]
Just want to get back here, have had zero bursts. So sounds like its fixed.

**Comment by Krzmbrzl** [comment]
Thanks for the feedback! At least we now have a hint as to where this issue is coming from.

It seems that it will be fixed either by fixing the instructions sent by the server or by finally ditching Celt...

**Comment by vimpostor** [comment]
So why are we even still supporting Celt? Opus is clearly far superior and I don't understand why Mumble still burdens itself by supporting Celt.

**Comment by Krzmbrzl** [comment]
I think the only reason is because nobody has yet taken on the task of stripping all Celt-related code out of Mumble. I think this can be a bit tedious since one has to make sure that all parts that are no longer needed are removed but the bits that are shared between Opus and Celt remain where they are (though a rename might be necessary for some things) :shrug:

**Comment by davidebeatrici** [comment]
Let's do it for 1.5.0.

**Comment by Krzmbrzl** [comment]
Well I am certainly not opposed to this :+1:

**Comment by TredwellGit** [comment]
How about setting `opusthreshold=0` and `iOpusThreshold = 0` for 1.4.0?

**Comment by Krzmbrzl** [comment]
I always assumed these were the defaults anyway. If not then yeah changing these sounds good to me. Do you want to create a PR @TredwellGit ?

**Comment by TredwellGit** [comment]
I will submit a pull request.

**Comment by RayneYoruka** [comment]
Has this been applied in to the RC of 1.5.xx? (The burst of noise are getting really loud)

**Comment by Krzmbrzl** [comment]
Yes
